R/utilities.R

```diff
@@ -1299,20 +1298,23 @@ coords=%s'
   cat(config, file=file.path(dir, "cellbrowser.conf"))
   message("Prepared cellbrowser directory ", dir)
 
-  if (!is.null(cb.dir) && require(reticulate)) {
-    library(reticulate)
+  if (!is.null(cb.dir)) {
     if (!py_module_available("cellbrowser")) {
-      py_install("cellbrowser") # see https://rstudio.github.io/reticulate/articles/python_packages.html
+      stop(
+        "Python package `cellbrowser` is required to prepare and run ",
+        "Cellbrowser. Please, install it with `pip install cellbrowser`."
```

Comment:
This may work for many people. It would be great if this was always correct, but unfortunately installing modules in python can be quite complicated, depending on the OS. Reticulate uses the first python on the path, which may be python3. If the OS has a pip, and pip3 and a pip2, then typing "pip" may install cellbrowser, but not into the python that reticulate uses.  So I would suggest adding another option, e.g. do.install that takes care of the install by running py_install("cellbrowser"). In addition, we could remind the user that they can always do require(reticulate) + use_python(pathname) to use a particular Python version that they like or a virtual env in that python with use_virtualenv("myenv") or a conda environment with use_condaenv(name). I know this is getting quite complicated, but it is likely that we have users here that are not Python fans or alternatively may have Python fans here that don't understand R well, and few people know reticulate, so I think it's good to make this part as easy as possible for them. If they stumble here, they may give up entirely...

Reply:
Hi Maximilian,

Reticulate facilitates using alternative Python installations via [`use_python`](https://www.rdocumentation.org/packages/reticulate/versions/1.10/topics/use_python), so if a user uses an alternative Python they can set reticulate to use that particular version of Python. In our experience, reticulate can switch between CPython 2.X and 3.X with relative ease, so we have little concern for pip installations. We have also noticed that reticulate has issues with Anaconda Python, so we encourage our users to use [CPython over Anaconda as well as use the system's version of Pip over conda](https://github.com/satijalab/seurat/issues/486#issuecomment-391457067) when using a Seurat function that calls out to Python. Using CPython and Pip installed from the system package manager or from [python.org](https://www.python.org) to install modules generally allow Seurat and reticulate to work without any extra effort on the user's end.

Moreover, CRAN generally frowns upon a package modifying the system outside of using temporary files. As per [CRAN's policies](https://cran.r-project.org/web/packages/policies.html):

> The code and examples provided in a package should never do anything which might be regarded as malicious or anti-social. The following are illustrative examples from past experience.
> - [...]
> - Packages should not write in the user’s home filespace (including clipboards), nor anywhere else on the file system apart from the R session’s temporary directory (or during installation in the location pointed to by TMPDIR: and such usage should be cleaned up). Installing into the system’s R installation (e.g., scripts to its bin directory) is not allowed.

As such, we are hesitant to include code that messes with a user's system, such as installing Python packages, and will instead help our users use CPython and Pip to install Python modules used by Seurat. Users savvy enough to use virtual or conda environments should be able to look through reticulate's documentation and figure out how to configure reticulate to work with their Python setup.

Reply:
This makes sense, I didn't know this CRAN rule. I'll remove the py_install call. thanks!